src/Foundation/NSObject2.cs

```diff
@@ -829,12 +829,18 @@ internal void ClearHandle ()
 			}
 		}
 
-		unsafe void FreeData ()
+		void FreeData ()
 		{
 			if (super != IntPtr.Zero) {
 				Marshal.FreeHGlobal (super);
 				super = IntPtr.Zero;
 			}
+#if NET
+			foreach (var field in GetType ().GetFields (BindingFlags.Instance | BindingFlags.NonPublic)) {
+				if (field.Name.StartsWith ("__mt_", StringComparison.Ordinal))
+					field.SetValue (this, null);
```

Comment:
To be honest, all these fields are causing leaks in VSMac, because they hold strong refs when the underlying native field is weakref.

The bindings exposing a strong ref property are part of the problem.

Reply:
Maybe having a BindingFlags.DeclaredOnly could limit the scope to private fields declared in xamarin.mac.